### ISSUE TITLE
Floating point exception when merging psf with srun -n 20 -c 2 desi_compute_psf_mpi

### ISSUE THREAD

**Issue by julienguy** [desihub/desispec#835]
Using specex master as installed at /global/common/software/desi/cori/desiconda/20190804-1.3.0-spec/code/specex/master:
```
time srun -n 20 -c 2 desi_compute_psf_mpi \
     --input-image /global/project/projectdirs/desi/spectro/redux/sjbailey/preproc/20191024/00020662/preproc-b3-00020662.fits \
     --input-psf /global/project/projectdirs/desi/spectro/desi_spectro_calib/trunk/spec/sp3/psf-sm4-b-science-slit-20191015.fits \
     --output-psf /global/project/projectdirs/desi/spectro/redux/sjbailey/exposures/20191024/00020662/psf-b3-00020662.fits
...
INFO wrote PSF in /global/project/projectdirs/desi/spectro/redux/sjbailey/exposures/20191024/00020662/psf-b3-00020662_11.fits
INFO:specex.py:260:merge_psf: Will merge 20 PSFs in /global/project/projectdirs/desi/spectro/redux/sjbailey/exposures/20191024/00020662/psf-b3-00020662.fits
INFO:specex.py:265:merge_psf: merging /global/project/projectdirs/desi/spectro/redux/sjbailey/exposures/20191024/00020662/psf-b3-00020662_01.fits into /global/project/projectdirs/desi/spectro/redux/sjbailey/exposures/20191024/00020662/psf-b3-00020662_00.fits
srun: error: nid00009: task 0: Floating point exception
srun: Terminating job step 25399083.1
slurmstepd: error: *** STEP 25399083.1 ON nid00009 CANCELLED AT 2019-10-24T20:52:04 ***
srun: error: nid00009: tasks 1-19: Terminated
srun: Force Terminated job step 25399083.1
```

This is not a specex issue, it's on the python side when running in the MPI environment with the desi_compute_psf_mpi script because the following runs just fine:
```
desi_merge_psf -i /global/project/projectdirs/desi/spectro/redux/sjbailey/exposures/20191024/00020662/psf-b3-00020662_*.fits -o $SCRATCH/psf.fits
```

**Comment by tskisner** [comment]
Random floating point exceptions got me thinking...  specex is the one place in our stack that uses BLAS/LAPACK functions in compiled code.  We are using MKL for that:
```
ldd $(which specex_merge_psf) 
        <snip>
	libmkl_rt.so => /opt/intel/compilers_and_libraries_2019.3.199/linux/mkl/lib/intel64/libmkl_rt.so (0x00002aaaac229000)
         ...
```
In our CMB stack at NERSC, we have seen silent memory corruption, segfaults, etc in a specific scenario:

1.  Numpy built with anaconda provided MKL.

2.  A compiled python extension which uses a different MKL provided by Intel compilers.

Basically the two MKL versions have some incompatibility, and only one can be loaded in a process.  Which version gets loaded depends on whether "import numpy" comes before or after the import of the compiled extension.  The two options that we have found that work are:

1.  Build compiled code with Intel compilers and MKL.  For python, use conda-forge packages *without* MKL.

2.  Build compiled code with gcc and OpenBLAS.  For python use packages that have MKL.

In our CMB stack, most of our heavy computational work is in compiled code, so we went with option (1).  For desispec, I think only specex is using much heavy compiled code, so option (2) would make more sense.

Again, I have no direct evidence that we are hitting this issue, but it is "suspicious" based on things we have seen with that other stack.  This MKL conflict only started happening with recent versions of the Intel compiler at NERSC and recent versions of conda packages.  It may be that in the past we were "lucky" and the two MKL versions were interoperable. 

I'll install a gcc-based desiconda on cori so that we could at least test whether this is the issue.

**Comment by julienguy** [comment]
Remember all bundle psfs are written to disk. So the psf fit was successful for all ranks.

**Comment by tskisner** [comment]
Yes- it may not explain this specific issue.  It is a more general problem.  I should actually go put this info into a desiconda issue.  Will do that now.

**Comment by sbailey** [comment]
Commentary -- the workflow is basically:
  1. individual ranks write one file per bundle using `libspecex.cspecex_desi_psf_fit()`
  2. MPI allreduce to check for failures (this should sync up all the ranks)
  3. rank 0 re-reads the individual files and writes out the merged file

Empirical hack: inserting a 5 sec pause between (2) and (3) makes the floating point exception go away.  It acts as if the files written by other ranks haven't really been closed and flushed fully before getting to (3), despite the allreduce barrier in (2).

Is specex perhaps relying upon some destructor to handle file closing that isn't getting triggered by `libspecex.cspecex_desi_psf_fit()` returning?  It looks like `specex::write_psf_fits` is calling `harp::fits::close(fp)` properly.

`<FUD>`Otherwise, if we are absolutely confident that specex is closing all of its files cleanly, then we may be hitting another "feature" of our favorite parallel file systems (in this case I've been testing on /project, not $SCRATCH).`</FUD>`

**Comment by sbailey** [comment]
One more data point: even within `desi_compute_psf_mpi` without the 5 sec timeout, the floating point exception isn't reproducible, so it is also possible that I was just "lucky" when I had that pause.  Either way, a magic 5 sec pause is not a good solution even if it temporarily helps a bit.

**Comment by tskisner** [comment]
I know it is a long shot, but it might be interesting to try with this software stack:
```
%>  module use /global/common/software/desi/users/kisner/cori/20191025-1.3.1-spec/modulefiles
%>  module load desi
```
This is desiconda built with gcc plus the master versions of all desi packages as of a couple hours ago.

**Comment by julienguy** [comment]
Did not help (was a bit random ...)
